ui/src/pages/Create.js

```diff
@@ -0,0 +1,645 @@
+          onChange={(e) => this.setState({formPhotoFile: e.target.files[0]})} />
+        <Button
+          className='pf-button-primary'
+          type='button'
+          onClick={() => this.photoUploadInput.current.click()}
+          disabled={this.state.formPhotoUrl != ""}>
+          {this.props.intl.formatMessage(messages.uploadPhoto)}
+        </Button>
+        {photoFilenameSpan}
+        {photoDeleteButton}
+        <TextField
+          label={this.props.intl.formatMessage(messages.orEnterPhotoUrl)}
+          outlined
+        >
+          <Input
+            value={this.state.formPhotoUrl}
+            onChange={(e) => this.setState({formPhotoUrl: e.target.value})}
+            disabled={this.state.formPhotoFile != ""} />
+        </TextField>
+      </PFNotchedOutline>
+    );
+  }
+
+  /*
+   * Adds a profile page field for the given site to the form.
+   */
+  addProfilePageField(site) {
+    this.setState({
+      formProfilePages: this.state.formProfilePages.concat(new Map(
+          [['site', site], ['value', '']])),
+      showProfilePageOptions: false,
+    });
+  }
+
+  /*
+   * Removes the given profile page field from the form.
+   */
+  removeProfilePageField(index) {
+    var newProfileList = this.state.formProfilePages.slice(0, index);
+    if (index < this.state.formProfilePages.length - 1) {
+      newProfileList = newProfileList.concat(
+          this.state.formProfilePages.slice(
+              index + 1, this.state.formProfilePages.length));
+    }
+    this.setState({
+      formProfilePages: newProfileList,
+    });
+  }
+
+  /*
+   * Renders a profile field menu option for the given site.
+   */
+  renderProfilePageOption(site) {
+    // TODO(nworden): include site icons
+    return (
+      <li
+        className='mdc-typography--body1'
+        onClick={() => this.addProfilePageField(site)}>
+        <FormattedMessage {...CREATE_PROFILE_PAGE_SITES[site]} />
+      </li>
+    );
+  }
+
+  /*
+   * Render the profile page link fields (if any) and the "Add site" button.
+   */
+  renderProfilePageFields() {
+    var profilePageFields = this.state.formProfilePages.map((page, index) => (
+      <div
+        className='create-profilefieldwrap'
+        key={this.profileFieldIdCounter++}>
+        <TextField
+          label={this.props.intl.formatMessage(
+              CREATE_PROFILE_PAGE_SITES[page.get('site')])}
+          outlined
+        >
+          <Input
+            value={this.state.formProfilePages[index].value}
+            onChange={(e) => {
+              // A component's state shouldn't be mutated directly. We need to
+              // update a map in an array, so we copy the map and do some
+              // splicing to produce a new array with the new map.
+              var newPageMap = new Map(this.state.formProfilePages[index]);
+              newPageMap['value'] = e.target.value;
+              const newPagesArr = this.state.formProfilePages.slice(0, index)
```

Comment:
Maybe consider using Immutable.js which makes this kind of operations easier?:
https://facebook.github.io/immutable-js/docs/#/List
It provides immutable `set` method (and `delete`, `inset` as well).

Reply:
Nice, that makes it much cleaner!